src/app/codegen-data-model-provider/EmberDataBuffer.cpp

```diff
@@ -0,0 +1,358 @@
+    int64_t value;
+
+    if (reader.GetType() == TLV::kTLVType_Null)
+    {
+        // we know mIsNullable due to the check at the top of ::Decode
+
+        // Most negative integer (i.e. 0b1000...0 is flagged as NULL value)
+        value = info.minValue;
+    }
+    else
+    {
+        ReturnErrorOnFailure(reader.Get(value));
+
+        // NULLABLE reserves minValue for NULL, so range is:
+        //   - NULLABLE:      (minValue, MaxValue]
+        //   - NON-NULLABLE:  [minValue, MaxValue]
+        bool valid = (value <= info.maxValue) && (mIsNullable ? (value > info.minValue) : (value >= info.minValue));
+
+        VerifyOrReturnError(valid, OutOfRangeError(info.byteCount));
+    }
+    writer.EndianPutSigned(value, info.byteCount);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EmberAttributeBuffer::DecodeAsString(chip::TLV::TLVReader & reader, PascalString stringType, TLV::TLVType tlvType,
+                                                EndianWriter & writer)
+{
+    // Handle null first, then the actual data
+    if (reader.GetType() == TLV::kTLVType_Null)
+    {
+        // we know mIsNullable due to the check at the top of ::Decode
+        switch (stringType)
+        {
+        case PascalString::kShort:
+            writer.Put8(0xFF);
+            break;
+        case PascalString::kLong:
+            writer.Put16(0xFFFF);
+            break;
+        }
+        return CHIP_NO_ERROR;
+    }
+
+    const uint32_t stringLength = reader.GetLength();
+
+    VerifyOrReturnError(reader.GetType() == tlvType, CHIP_ERROR_WRONG_TLV_TYPE);
+    VerifyOrReturnError(reader.GetLength() <= MaxLength(stringType, mIsNullable), CHIP_ERROR_INVALID_ARGUMENT);
+
+    // Size is a prefix, where 0xFF/0xFFFF is the null marker (if applicable)
+    switch (stringType)
+    {
+    case PascalString::kShort:
+        writer.Put8(static_cast<uint8_t>(stringLength));
+        break;
+    case PascalString::kLong:
+        writer.Put16(static_cast<uint16_t>(stringLength));
+        break;
+    }
+
+    // data copy
+    const uint8_t * tlvData;
+    ReturnErrorOnFailure(reader.GetDataPtr(tlvData));
+    writer.Put(tlvData, stringLength);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR EmberAttributeBuffer::Decode(chip::TLV::TLVReader & reader)
+{
+    // all methods below assume that nullable setting matches  (this is to reduce code size
+    // even though clarity suffers)
+    VerifyOrReturnError(mIsNullable || reader.GetType() != TLV::kTLVType_Null, CHIP_ERROR_WRONG_TLV_TYPE);
+
+    EndianWriter endianWriter(mDataBuffer.data(), mDataBuffer.size());
+
+    switch (mAttributeType)
+    {
+    case ZCL_BOOLEAN_ATTRIBUTE_TYPE: // Boolean
+        // Boolean values:
+        //   0x00 is FALSE
+        //   0x01 is TRUE
+        //   0xFF is NULL
+        if (reader.GetType() == TLV::kTLVType_Null)
+        {
+            // we know mIsNullable due to the check at the top of ::Decode
+            endianWriter.Put8(0xFF);
```

Comment:
Because this is an exercise in reducing code. I am explicitly trying to stay away from templates unless they are super readable and clearly small.

For null value maybe it is ok, I will have to check. For many things (like max size) for example the complexity overhead is too large.

Reply:
Updated some 0xFF / 0xFFFF with kNullValue constants for bools and string lengths.

Left the min/max alone as the existing templates require isNullable arguments and they add to much readability overhead.